### ISSUE TITLE
[$250] [HOLD for payment 2024-06-11] [Wave Collect] [Xero] [Imported Features and Taxes] Make Categories, Tags, and Taxes read-only when the user is connected to Xero

### ISSUE THREAD

**Issue by lakchote** [Expensify/App#39911]
As noted [here](https://github.com/Expensify/App/issues/39724) for QBO, we need to prevent the user from taking actions when the user is connected to Xero.

When the user connects to Xero:

- optimistically clear all existing categories, tags and taxes the user might have had on the policy, when they add the Xero connection
- prevent the user from turning off categories, tags and taxes features
- prevent the user from deleting categories, tags and taxes imported from Xero
- prevent the user from adding new categories, tags and taxes

This will be worked on by SWM engineers, see Slack discussion [here](https://expensify.slack.com/archives/C04878MDF34/p1712333484848839). 

Until the Xero authorization flow ([NewDot issue](https://github.com/Expensify/App/issues/39725))  is done (which entails NewDot, Web-E and IS changes), we won't be able to work on this yet.

<details><summary>Issue Owner</summary>Current Issue Owner: @</details>
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01bf513d396cfd4d35</li>
        <li>Upwork Job ID: 1800169199603790711</li>
        <li>Last Price Increase: 2024-06-10</li>
    </ul>
</details>

**Comment by joekaufmanexpensify** [comment]
@lakchote should I be finding someone from SWM to do this, or just waiting for someone who already committed to do it to pick it up?

**Comment by lakchote** [comment]
> @lakchote should I be finding someone from SWM to do this, or just waiting for someone who already committed to do it to pick it up?

I've asked [here](https://expensify.slack.com/archives/C04878MDF34/p1713196095864759?thread_ts=1712333484.848839&cid=C04878MDF34), if SWM engineers can pick this up!

**Comment by SzymczakJ** [comment]
Hey! I’m Jakub Szymczak from Software Mansion, an expert agency, and I’d like to work on this issue!

**Comment by joekaufmanexpensify** [comment]
Great. TY! @SzymczakJ is there an eta for a PR here?

**Comment by SzymczakJ** [comment]
@joekaufmanexpensify there should be the same task for QBO integration that would solve this issue(they also connect to accounting software and should make categories etc. read-only), but I cannot find it, so I'm not sure if it's already worked on by them. Also if nobody is taking care of it I do it 😃 cc @mountiny

**Comment by joekaufmanexpensify** [comment]
@SzymczakJ, are you referring to [this issue](https://github.com/Expensify/App/issues/39724)? If so, do you mean that the QBO issue will solve this for Xero too, and we can consolidate both into that issue?

**Comment by SzymczakJ** [comment]
Sorry for late response, I was OOO. The issue you mentioned isn't solving this issue completely, it only disables switches for taxes/categories but the user is still able to edit categories etc.(as shown on video). 

https://github.com/Expensify/App/assets/88395093/6ca56ed4-3e99-4c6c-b2d2-c0d2684ffdb8

Previously I thought that making categories/taxes read-only was also the case for QBO and that's why I said it will be solved with QBO but maybe that isn't the case. **Anyways the issue you mention isn't solving our issue completely and I don't see any open QBO issue that would solve this, so I could take care of that.** cc @lakchote 

Also is it only a Xero thing that the categories should be read-only and in case of QBO the user should be able to edit it freely?

**Comment by lakchote** [comment]
> Previously I thought that making categories/taxes read-only was also the case for QBO and that's why I said it will be solved with QBO but maybe that isn't the case. **Anyways the issue you mention isn't solving our issue completely and I don't see any open QBO issue that would solve this, so I could take care of that.** cc @lakchote
> 
> Also is it only a Xero thing that the categories should be read-only and in case of QBO the user should be able to edit it freely?

It shouldn't be a Xero thing only; from my understanding when you have an accounting integration connected it is the source of truth and the categories should come from this source only. So, it should be read only for both QBO and Xero from my understanding.

Could we have your opinions @trjExpensify @zanyrenney please?

**Comment by SzymczakJ** [comment]
Another question from design/product side: How read-only state should be displayed? I have some idea but I want to make sure.
This is how edit category flow looks like

https://github.com/Expensify/App/assets/88395093/7813d2ce-c79a-404e-afee-0eb80db5bc89

**We have few options of editing:
a) enable/disable category with a toggle click
b) change the name of category with clicking on category name
c) delete category by clicking three dots
d) select multiple categories and disable/delete them**

My idea would look like this: 
We are disabling enable/disable toggle( **a) option**), making category name MenuItem non-interactive, so only the name is displayed ( **b) option**), deleting three dots on CategorySettingsPage ( **c) option**) and finally disabling select multiple options on category list on WorkspaceCategoriesPage ( **d) option**).


https://github.com/Expensify/App/assets/88395093/ce890d7e-8e71-413f-970a-b33fb35c67bf

WDYT @zanyrenney @trjExpensify

**Comment by SzymczakJ** [comment]
another question: the user has ability to enable/disable import of taxes,categories, etc.. We should only make taxes(and similarly categories, etc.) read-only when the import taxes toggle is on, right?

**Comment by trjExpensify** [comment]
>1. You can still add/delete tags when being connected to accounting integration
>2. You can still add/ delete taxes when being connected to accounting integration

Did you test on dev/main, yeah? If so, @hayata-suenaga what did your PR for https://github.com/Expensify/App/issues/41274 implement? 

>Should taxes, etc. be read-only when we are connected to accounting integration(with or without clicking import) or when we are connected to accounting integration AND clicking on import taxes, etc

Can you rephrase the question, I'm not sure I'm following you.

**Comment by SzymczakJ** [comment]
> Did you test on dev/main
I was testing on main 😄 

To make the question clearer I'll focus on categories.
Categories should be read-only when:
a) we are connected to any accounting integration
b) we are connected to accounting integration and have turned on the Chart of accounts toggle as on a screenshot 
<img width="554" alt="Screenshot 2024-05-07 at 15 23 43" src="https://github.com/Expensify/App/assets/88395093/724bcf1b-649d-4307-a5fc-bfdb5bbaf619">

Which of these options is correct? Right now we are using option **a**, but option **b** seems more logical to me and that's why i'm asking.

**Comment by trjExpensify** [comment]
Yeah, so that toggle has nothing to do with importing the chart of accounts or not - you can't not import those when connected to an accounting solution, it's the baseline requirement for the integration and thus not a choice. We've changed the design of that page to make it clearer as well: 

![image](https://github.com/Expensify/App/assets/16232057/b85cb949-ed1d-4f39-910a-607ded5abe6e)

**Comment by trjExpensify** [comment]
For context on that second toggle:

It's a different feature that gives you the ability to decide whether you want newly imported chart of accounts to be enabled or disabled by default when imported. So if you add a new account in QBO/Xero > sync the connection to import it into Expensify > the category in the table of Categories will be either enabled or disabled depending on this setting preference. If a category is disabled, it doesn’t show on someone’s expense to select when they create an expense on the workspace. If it's enabled, it does show in the list on an expense to select.

This feature exists because some integrations don't let you just import expense accounts as categories. So some customers prefer to have them imported as disabled by default to avoid showing a bunch of accounts to their members that aren’t intended to every be used to code their expenses. Once imported, they manually enable the category in the categories table to have it shown to members.

**Comment by SzymczakJ** [comment]
Oh that makes sense now.

And what about tags and taxes?
Should I treat it the same way as categories are treated(import is a baseline requirement)? Or should I check if the user has turned on some toggles like this QBO classes import.
<img width="334" alt="Screenshot 2024-05-07 at 15 54 01" src="https://github.com/Expensify/App/assets/88395093/66940e67-2438-4543-8ed3-d91e43b8c739">

**Comment by trjExpensify** [comment]
Only chart of accounts are mandatory for import, other coding is entirely optional. 

What are you trying to do though? I'm still confused why you'd be locking the import toggle on additional Xero coding like Tracking Categories and Taxes. 🤔

**Comment by SzymczakJ** [comment]
Sorry for the confusion 😅 . I'm not trying to lock the toggle for importing Taxes etc.. What I want to to is to determine if Tags should be read-only based on user switching one or more of these toggles(QBO integration in this case)
<img width="335" alt="image" src="https://github.com/Expensify/App/assets/88395093/697b557e-f673-41ec-b999-0375f69089eb">
<img width="333" alt="Screenshot 2024-05-07 at 16 18 16" src="https://github.com/Expensify/App/assets/88395093/14e452d3-2986-4cb4-abbe-f997cffe0165">
<img width="442" alt="image" src="https://github.com/Expensify/App/assets/88395093/effeace4-bc69-4c70-afb2-379548c8b4b6">
is that correct approach? right now we are looking only on the fact if any of the integrations is connected

**Comment by trjExpensify** [comment]
Haha all good, we'll get there! 😄 Let's maybe unpack "read-only" to ensure we're on the same page, and go from there. 

**- Can you turn on/off the `Categories`, `Tags` or `Taxes` feature in `More features` when connected to an accounting solution?** 

No, you can't. Once connected to an accounting solution, the only way `Categories`, `Tags` or `Taxes` can be enabled or disabled as features is by importing them via your connected accounting solution settings on the `Import` page. 

**- Can you freely `Add` or `Delete` a new category, tag or tax value when connected to an accounting solution?** 

No, you can't. If you want to add or delete an imported value you have to do that in the accounting solution as it's the "source of truth" for coding values.

**- Can you `enable` or `disable` a category, tag or tax value when connected to an accounting solution, such that the value does or doesn't show when someone creates an expense on the workspace?** 

Yes, you can. You can enable/disable individual imported values. There's an exception irrespective of accounting connections that doesn't allow you to disable a tax value if it's being used as one of the workspace currency defaults.

**Comment by SzymczakJ** [comment]
Nice, now I understand.
So basically that's answer to all of my questions and almost everything is done 😄 
 
The only thing left is to take care of these issues that I've pointed out, has anybody taken care of it already?
> I've looked into this and found couple things that didn't get solved.
> 
> 1. You can still add/delete tags when being connected to accounting integration
> 
> <img alt="Screenshot 2024-05-07 at 14 56 40" width="1531" src="https://private-user-images.githubusercontent.com/88395093/328529653-06541040-39f8-49e1-ad86-b4cc7dc3f57a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUwOTA5MDYsIm5iZiI6MTcxNTA5MDYwNiwicGF0aCI6Ii84ODM5NTA5My8zMjg1Mjk2NTMtMDY1NDEwNDAtMzlmOC00OWUxLWFkODYtYjRjYzdkYzNmNTdhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA3VDE0MDMyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYwMTE5NGFmYTllMzkwYmQxNzEzOGViYWViODVhZWI3MThjZjFjODZkYmIwYzJiZjgyMDMxNzcwZmNlYzdmZDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.TS8M7Wv2GD0Av-GxFwyNa6qUVNX6vxXaQfe2SYKCn7c"> 2. You can still add/ delete taxes when being connected to accounting integration <img alt="Screenshot 2024-05-07 at 14 59 06" width="1535" src="https://private-user-images.githubusercontent.com/88395093/328530447-ab26959f-86ba-4e8c-a416-f1a15e1be39a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUwOTA5MDYsIm5iZiI6MTcxNTA5MDYwNiwicGF0aCI6Ii84ODM5NTA5My8zMjg1MzA0NDctYWIyNjk1OWYtODZiYS00ZThjLWE0MTYtZjFhMTVlMWJlMzlhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTA3VDE0MDMyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdjZWI3Nzk3MTNhMDRiZWU0ODc3OWYzMTVkNzAzNTA5MzRjNDMzZGM5OTQ4NTQ2NGMyMDQzODE2MGZjMDFhNTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.9LnZqLDZoto6lOFqOenl0NDP-kC-9uOv1Loqgya5nNM"> I could fix that issues in this task.

**Comment by trjExpensify** [comment]
Nice! So to confirm, the PR for this issue https://github.com/Expensify/App/issues/41274 didn't take care of not allowing someone to `Add` or `Delete` tax values. Is that right?

**Comment by SzymczakJ** [comment]
Correct

**Comment by trjExpensify** [comment]
Cool okay, so @hayata-suenaga's PR must have missed that and only covered `Categories` and `Tags`. Please do apply the same logic the tax values on the `Taxes` page as well. :)

**Comment by SzymczakJ** [comment]
This fix is really small and I've already pushed out a PR with it. I will fill PR author checklist tomorrow. When we merge it we can count this issue as completed 😄

**Comment by SzymczakJ** [comment]
I've found bug related to importing taxes.
When we choose to import taxes in accounting integration page then policy(specifically `policy?.tax?.trackingEnabled` prop) is not updated on backend(I suppose it's on the backend side, because when importing tags everything works fine and policy is properly updated). I've checked GH and I don't think we have any issue tracking that.
**So in order to test taxes you have to connect to QBO, turn on taxes on more features page and then import taxes on QBO import taxes page**
@lakchote @trjExpensify

**Comment by trjExpensify** [comment]
There was this PR of @aldo-expensify's for fixing the import of QBO taxes: https://github.com/Expensify/App/pull/41283

The QA steps below and video example on staging [here](https://github.com/Expensify/App/pull/41283#issuecomment-2089564082):

> 1. Create account in NewDot
> 2. Create workspace
> 3. Enable "Accounting" in "More Features"
> 4. Connect the workspace to QBO
> 5. Go to "Import" and enable "Taxes"
> 6. Click the 3 vertical dots and click "Sync now"
> 7. See that the "Taxes" menu entry appears in the left hand side and taxes were imported in it

Then it looks like @hayata-suenaga has this [Web-E PR](https://github.com/Expensify/Web-Expensify/pull/41897) to "Trigger sync between New Expensify and QBO when any of the configuration changes", which presumably removes the need for steps 6&7 to give it a manual push? 

>because when importing tags everything works fine and policy is properly updated).

Interestingly, I haven't been able to get Tags to work with the Xero integration at all. There's a known issue with QBO when it comes to rendering multiple levels of tags [here](https://github.com/Expensify/App/pull/41351).

**Comment by SzymczakJ** [comment]
Okay it works with QBO "sync now" button click, but sync now button doesn't work for Xero. I guess when @hayata-suenaga PR will be merged and the same will be done for Xero, then everything will work.



> Interestingly, I haven't been able to get Tags to work with the Xero integration at all.

I was testing tags on an old workspace and that's why it was working for me 😅 I think we could test it that way because it's not about importing right tags/taxes but about making taxes read-only, so as long as we have anything on tax/tags page and it's read-only then we are good.

**Comment by trjExpensify** [comment]
>Okay it works with QBO "sync now" button click, but sync now button doesn't work for Xero.

I think for sync now to work with Xero, we're waiting on this PR actually: https://github.com/Expensify/App/pull/41562 - the corresponding Web-E PR hit staging earlier. 

>I guess when @hayata-suenaga PR will be merged and the same will be done for Xero, then everything will work.

Maybe @lakchote can confirm, but if the triggering of the sync when a Xero setting is changed isn't being taken care of somewhere else - we could [add it to @hayata-suenaga's PR](https://github.com/Expensify/Web-Expensify/pull/41897/files) so we get both done in that Web-E PR.

**Comment by lakchote** [comment]
Yes, `Sync now` currently doesn't work for Xero, since we need this app [PR](https://github.com/Expensify/App/pull/41562). I've pinged the contributor on [Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1715166591695889?thread_ts=1714119216.357459&cid=C01GTK53T8Q) to make sure we're moving forward.

**Comment by SzymczakJ** [comment]
So maybe let's wait for Xero import features to be merged so we can reliably test this. WDYT @trjExpensify @lakchote

**Comment by trjExpensify** [comment]
Sounds good!

**Comment by hayata-suenaga** [comment]
> I do agree. Let's do both in this [PR](https://github.com/Expensify/Web-Expensify/pull/41897/files), it shouldn't need too much changes and we'll avoid back and forth that way.

@lakchote, the linked App PR is triggering sync on the frontend side while my existing Web-E PR does so in the backend. Which pattern should we use? 😄 More detailed comment in the App PR your contributor is working on -> https://github.com/Expensify/App/pull/41562/files#r1594428819

**Comment by trjExpensify** [comment]
(Ignore me)

https://github.com/Expensify/App/pull/41785 in review.

**Comment by lakchote** [comment]
> @lakchote, the linked App PR is triggering sync on the frontend side while my existing Web-E PR does so in the backend. Which pattern should we use? 😄 More detailed comment in the App PR your contributor is working on -> https://github.com/Expensify/App/pull/41562/files#r1594428819

Answere you [here](https://github.com/Expensify/App/pull/41562/files#r1595196874), we're triggering the manual sync from the frontend when clicking on `Sync now`, but the sync is done in the backend using the [SyncPolicyToXero](https://github.com/Expensify/Web-Expensify/pull/41875) API command. Let me know if I've missed something 😅

**Comment by hayata-suenaga** [comment]
> Answered you [here](https://github.com/Expensify/App/pull/41562/files#r1595196874), we're triggering the manual sync from the frontend when clicking on Sync now, but the sync is done in the backend using the https://github.com/Expensify/Web-Expensify/pull/41875 API command. Let me know if I've missed something 😅

thanks! I seem to have misunderstood the code. Please ignore my previous comments 😓

**Comment by joekaufmanexpensify** [comment]
PR in review

**Comment by SzymczakJ** [comment]
@trjExpensify that's a quick one-liner fix. Sorry for missing that 😄

**Comment by trjExpensify** [comment]
All good, thanks! Assigning @hungvu193 to review.

**Comment by joekaufmanexpensify** [comment]
PR on staging

**Comment by joekaufmanexpensify** [comment]
@SzymczakJ is with SWM, so only payment needed here is $250 to @hungvu193 for C+ review via Upwork.

**Comment by hungvu193** [comment]
Hey @joekaufmanexpensify , I got paid already so no payment needed here 😄

**Comment by joekaufmanexpensify** [comment]
Ah, okay, sounds good! Just withdrew the upwork offer and closed the upwork job.

**Comment by joekaufmanexpensify** [comment]
In that case, we should be all set here then. Thanks everyone!
